File: ruby/day11_part2_2018.rb
serial = File.read('input.txt').strip.to_i

grid_size = 300
grid = Array.new(grid_size) { Array.new(grid_size, 0) }

(0...grid_size).each do |y|
  (0...grid_size).each do |x|
    rack_id = x + 11
    power_level = rack_id * (y + 1)
    power_level += serial
    power_level *= rack_id
    power_level = (power_level / 100) % 10
    power_level -= 5
    grid[y][x] = power_level
  end
end

max_power = -Float::INFINITY
max_x = max_y = max_size = 0

(1..grid_size).each do |size|
  break if size > 20  # Limit the maximum size to 20

  (0..grid_size - size).each do |y|
    (0..grid_size - size).each do |x|
      total_power = (0...size).sum { |dy| (0...size).sum { |dx| grid[y + dy][x + dx] } }
      if total_power > max_power
        max_power = total_power
        max_x, max_y, max_size = x + 1, y + 1, size
      end
    end
  end
end

puts "#{max_x},#{max_y},#{max_size}"

File: julia/day11_part2_2018.jl

function solve()
    serial = parse(Int, strip(read("input.txt", String)))
    gridSize = 300
    grid = zeros(Int, gridSize, gridSize)

    for y in 1:gridSize, x in 1:gridSize
        rackID = x + 10
        powerLevel = rackID * y
        powerLevel += serial
        powerLevel *= rackID
        powerLevel = (powerLevel ÷ 100) % 10
        powerLevel -= 5
        grid[y, x] = powerLevel
    end

    maxPower = -typemax(Int)
    maxX, maxY, maxSize = 0, 0, 0

    for size in 1:gridSize
        for y in 1:(gridSize - size + 1), x in 1:(gridSize - size + 1)
            totalPower = sum(grid[y:y+size-1, x:x+size-1])
            if totalPower > maxPower
                maxPower = totalPower
                maxX, maxY, maxSize = x, y, size
            end
        end
    end

    println("$(maxX),$(maxY),$(maxSize)")
end

solve()
